subsys/net/ip/tcp2.c

```diff
@@ -1275,6 +1279,110 @@ static enum net_verdict tcp_recv(struct net_conn *net_conn,
 	return NET_DROP;
 }
 
+static uint32_t seq_scale(uint32_t seq)
+{
+	return seq + (k_ticks_to_ns_floor32(k_uptime_ticks()) >> 6);
+}
+
+static uint8_t secret_key[16]; /* MD5 128 bits as described in RFC6528 */
+
+static uint32_t tcpv6_init_isn(struct in6_addr *saddr,
+			       struct in6_addr *daddr,
+			       uint16_t sport,
+			       uint16_t dport)
+{
+	struct {
+		uint16_t secret_key[sizeof(secret_key)];
+		struct in6_addr saddr;
+		struct in6_addr daddr;
+		uint16_t sport;
+		uint16_t dport;
+	} buf = {
+		.saddr = *(struct in6_addr *)saddr,
+		.daddr = *(struct in6_addr *)daddr,
+		.sport = sport,
+		.dport = dport
+	};
+
+	uint8_t hash[16];
+	static bool done;
+
+	if (!done) {
+		if (IS_ENABLED(CONFIG_CSPRING_ENABLED)) {
+			sys_csrand_get(secret_key, sizeof(secret_key));
+		} else {
+			sys_rand_get(secret_key, sizeof(secret_key));
+		}
+
+		done = true;
+	}
+
+	memcpy(buf.secret_key, secret_key, sizeof(secret_key));
+	mbedtls_md5_ret((const unsigned char *)&buf, sizeof(buf), hash);
+
+	return seq_scale(hash[0]);
+}
+
+static uint32_t tcpv4_init_isn(struct in_addr *saddr,
+			       struct in_addr *daddr,
+			       uint16_t sport,
+			       uint16_t dport)
+{
+	struct {
+		uint16_t secret_key[sizeof(secret_key)];
+		struct in_addr saddr;
+		struct in_addr daddr;
+		uint16_t sport;
+		uint16_t dport;
+	} buf = {
+		.saddr = *(struct in_addr *)saddr,
+		.daddr = *(struct in_addr *)daddr,
+		.sport = sport,
+		.dport = dport
+	};
+
+	uint8_t hash[16];
+	static bool done;
+
+	if (!done) {
+		if (IS_ENABLED(CONFIG_CSPRING_ENABLED)) {
+			sys_csrand_get(secret_key, sizeof(secret_key));
```

Comment:
My only concern here is if this is slow enough to cause a delay creating initial connections. As long as the hash function is secure (which MD5 is not, unfortunately), it should be safe to use the same secret across multiple connections. However, I'm fine with this if it doesn't cause any undue performance issues.

Reply:
Linux uses siphash instead of md5. Unfortunately we do not have that in zephyr atm, but could use it later if needed.

Reply:
> My only concern here is if this is slow enough to cause a delay creating initial connections.

Note that this value is only calculated once, so should not be too big issue.

Reply:
Although it doesn't technically meet the requirements of the RFC, siphash isn't exactly any weaker than MD5. It's main concern from a security perspective is the small size, but that doesn't matter here as we are only getting 32 bits out of it. So, I'd be fine with using siphash as is.

Reply:
Yes, I agree. Beside RFC only propses MD5, it does not mandate that it must be used.
But this talk is cheap as we do not have siphash in zephyr. I do not have plans to add that myself, so wondering if someone from security team would be willing to contribute that?

Reply:
I find this weird to do a conditional on CSPRING... either you require it or you don't. It seems like you should just use the required rand() function.

Reply:
> To explain further, the random value being requested either needs to be random or cryptographically random.

It's actually in a kind of weird middle ground. The result is a 32-bit value, which clearly isn't large enough to really be considered cryptographically sound. The main property needed is unpredictability. With this RFC change, even using a non-crypto rand generator (such as sys_rand32) would be fine, as long as the initial seed can be made unpredictable. It could be cryptographically strong, or just something not visible from outside.

So, in this case, it might make sense to make use of the sys_rand32, as long as there is some aspect of it that is unknown externally.

Reply:
> To explain further, the random value being requested either needs to be random or cryptographically random.

Just as other folks, I don't think its either-or. Here we clearly need just (pseudo)random, but if we have cryptorandom around, we use it here too.

Maybe factor out that twice-repeating `if` as an inline func or macro, with such an explanatory comment.

Reply:
I disagree. Just because the system has a cryptorandom doesn't mean random needs should use it. I think of it as "self documenting code" in this situation as it becomes evident that the random value being requested needs to be cryptographically secure or it doesn't... If there is a thought to have rand() actually call csrand() if it exists then lets push that down to the definition of rand32 where there is a Kconfig to have it use the csrand()... lets not populate it up to the actual application code.

Also consider the performance impact difference (if any) of using csrand vs rand